addons/account_edi/models/account_edi.py

```diff
@@ -0,0 +1,251 @@
+
+        :param attachment:  An ir.attachment record.
+        :returns:           The invoice where to import data.
+        """
+        for file_data in self._decode_attachment(attachment):
+            for edi in self:
+                res = False
+                if file_data['type'] == 'xml':
+                    res = edi._update_invoice_from_xml_tree(file_data['filename'], file_data['xml_tree'], invoice)
+                elif file_data['type'] == 'pdf':
+                    res = edi._update_invoice_from_pdf_reader(file_data['filename'], file_data['pdf_reader'], invoice)
+                    file_data['pdf_reader'].stream.close()
+                if res:
+                    return res
+            return False
+
+    def _decode_attachment(self, attachment):
+        """Decodes an ir.attachment and unwrap sub-attachment into a list of dictionary each representing an attachment.
+
+        :param attachment:  An ir.attachment record.
+        :returns:           A list of dictionary for each attachment.
+        * filename:         The name of the attachment.
+        * content:          The content of the attachment.
+        * type:             The type of the attachment.
+        * xml_tree:         The tree of the xml if type is xml.
+        * pdf_reader:       The pdf_reader if type is pdf.
+        """
+        def _decode_xml(content):
+            try:
+                return etree.fromstring(content)
+            except Exception as e:
+                _logger.exception("Error when converting the xml content to etree: %s" % e)
+                return None
+
+        def _decode_pdf(content):
+            try:
+                buffer = io.BytesIO(content)
+                return OdooPdfFileReader(buffer)
+            except Exception as e:
+                # Malformed pdf
+                _logger.exception("Error when reading the pdf: %s" % e)
+                return None
+
+        content = base64.b64decode(attachment.with_context(bin_size=False).datas)
+        to_process = []
+
+        if 'pdf' in attachment.mimetype:
+            pdf_reader = _decode_pdf(content)
+            if pdf_reader:
+
+                # Process embedded files.
+                for filename, content in pdf_reader.getAttachments():
+                    xml_tree = _decode_xml(content)
+                    if len(xml_tree):
+                        to_process.append({
+                            'filename': filename,
+                            'content': content,
+                            'type': 'xml',
+                            'xml_tree': xml_tree,
+                        })
+
+                # Process the pdf itself.
+                to_process.append({
+                    'filename': attachment.name,
+                    'content': content,
+                    'type': 'pdf',
+                    'pdf_reader': pdf_reader,
+                })
+
+        elif 'xml' in attachment.mimetype:
```

Comment:
you sure if you should do anything, but I see you don't try to unwrap a single XML attachment into several ir.attachment.

So you might want to move out the inner function `_decode_xml` to allow some XML specific format to be unwrapped as well in several files (I think the EFFF can include the PDF inside the XML, some other yet unknown format might include several invoices in a single XML doc...).

What do you think?

Reply:
I moved the _decode functions outside so we'll be able to override if necessary.

As for EFFF, the pdf is inside the XML but right it doesn't seem necessary to process it, so we're not doing it for now.